appserver/payara-appserver-modules/microprofile/jwt-auth/src/main/java/fish/payara/microprofile/jwtauth/eesecurity/JwtPublicKeyStore.java

```diff
@@ -0,0 +1,297 @@
+        Optional<String> publicKey = readDefaultPublicKey();
+        
+        if (!publicKey.isPresent()) {
+            publicKey = readMPEmbeddedPublicKey();
+        }
+        if (!publicKey.isPresent()) {
+            publicKey = readMPPublicKeyFromLocation();
+        }
+        return publicKey;
+    }
+    
+    private Optional<String> readDefaultPublicKey() {
+        return readPublicKeyFromLocation("/publicKey.pem");
+    }
+
+    private Optional<String> readMPEmbeddedPublicKey() {
+        return config.getOptionalValue(VERIFIER_PUBLIC_KEY, String.class);
+    }
+
+    private Optional<String> readMPPublicKeyFromLocation() {
+        Optional<String> locationOpt = config.getOptionalValue(VERIFIER_PUBLIC_KEY_LOCATION, String.class);
+
+        if (!locationOpt.isPresent()) {
+            return Optional.empty();
+        }
+
+        String publicKeyLocation = locationOpt.get();
+
+        return readPublicKeyFromLocation(publicKeyLocation);
+    }
+    
+    private Optional<String> readPublicKeyFromLocation(String publicKeyLocation) {
+
+        URL publicKeyURL = currentThread().getContextClassLoader().getResource(publicKeyLocation);
+
+        if (publicKeyURL == null) {
+            try {
+                publicKeyURL = new URL(publicKeyLocation);
+            } catch (MalformedURLException ex) {
+                publicKeyURL = null;
+            }
+        }
+        if (publicKeyURL == null) {
+            return Optional.empty();
+        }
+        
+        try { 
+            return readPublicKeyFromURL(publicKeyURL);
+        } catch(IOException ex) {
+            throw new IllegalStateException("Failed to read public key.", ex);
+        }
+    }
+    
+    private Optional<String> readPublicKeyFromURL(URL publicKeyURL) throws IOException {
+        
+        URLConnection urlConnection = publicKeyURL.openConnection();
+        Charset charset = Charset.defaultCharset();
+        ContentType contentType = ContentType.newContentType(urlConnection.getContentType());
+        if(contentType != null) {
+            String charEncoding = contentType.getCharacterEncoding();
+            if(charEncoding != null) {
+                try {
+                    if (!Charset.isSupported(charEncoding)) {
+                        LOGGER.warning("Charset " + charEncoding + " for remote public key not supported, using default charset instead");
+                    } else {
+                        charset = Charset.forName(contentType.getCharacterEncoding());
+                    }
+                }catch (IllegalCharsetNameException ex){
+                    LOGGER.severe("Charset " + ex.getCharsetName() + " for remote public key not support, Cause: " + ex.getMessage());
+                }
+            }
+        }
+        try (InputStream inputStream = urlConnection.getInputStream();
+            BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, charset))){
```

Comment:
Does it make sense to read the HTTP cache headers here and use an eventually provided TTL for caching? Just as an override to the configured one.

Reply:
I like that. Assuming we keep the TTL configurable in `payara-mp-jwt.properties` does this priority sound ok for determining the effective TTL? The HTTP headers would be preferred unless the developer overrides it in the configuration.

1. TTL specified in `payara-mp-jwt.properties`. Disabled if set to 0.
2. Cache headers from an `HttpURLConnection`
3. Standard default of n minutes

Reply:
I wonder if its better to switch priorities 1 and 2. Thus honouring the HTTP header TTL the most although the configured TTL is set to 0. This allows the source of the public key to define the TTL which IMHO is the owner of the public key. Only if the source does not define any TTL the client/user of the public key may have it own rules for caching.

Reply:
My only hesitation with switching 1 and 2 was that the developer is powerless to override the cache headers. I'd hope that the HTTP cache headers are reasonable but we can't guarantee that will be the case. Perhaps unreasonable cache headers isn't a problem Payara should be trying to solve. I'm torn.

I'll get to work on this with the approach you suggested. I need to think through a good way to change the TTL at key load time. If we change our minds on priority, it should be an easy switch 🙂 .

Thanks for the feedback!

Reply:
Maybe just avoid unreasonable TTL header values which are too short lived or too long. Whatever this is. ;)

Reply:
I suppose "unreasonable" is pretty arbitrary 😅. I'll begin making those changes as soon as I can. I won't be as available with the holidays approaching but I should be able to push an update before the year ends.

Reply:
Sure. Happy holidays!

Reply:
@ghunteranderson, will you make the changes to consider the TTL in the cache headers of the response then?

Reply:
Hello @fturizo. Thanks for the feedback 🙂  I'll make these changes including the HTTP one. It's taken me a little bit longer than I expect to get back to this.

Reply:
I've added support for reading the max-age header and using that as an override for the cache TTL. It's ready for review whenever you are.